torch/_dynamo/variables/torch.py

```diff
@@ -773,13 +773,28 @@ def speculate_subgraph(
             output = f.call_function(tx, args, {})
             # Register output to graph
             # Modeled off of compile_and_call_fx_graph
-            # TODO: support non single Tensor output
+            # TODO: support pytree output
             # We check always_restore because we dont use the output or side effects of always_restore code,
             # like bwd.
-            if not isinstance(output, TensorVariable) and not always_restore:
-                unimplemented(
-                    "HigherOrderOperator with body with non single Tensor output"
-                )
+            if (
+                not isinstance(output, (TensorVariable, ListVariable, TupleVariable))
+                and not always_restore
+            ):
+                unimplemented("HigherOrderOperator with body with pytree output")
+
+            if isinstance(output, (ListVariable, TupleVariable)):
+
+                def _is_supported(var):
+                    return isinstance(var, TensorVariable) or (
+                        isinstance(var, ConstantVariable) and var.value is None
+                    )
+
+                if any(
+                    not _is_supported(var) for var in output.unpack_var_sequence(tx)
+                ):
+                    unimplemented(
+                        "HigherOrderOperator body's output must consist of tensors or None only"
+                    )
```

Comment:
Do we really need to support None output?

Reply:
yeah, the reason I added is because the backward logic of autograd.Function can consist of None and we don't want it to graph break. The specific example I encountered is this: https://github.com/pytorch/pytorch/blob/main/test/dynamo/test_repros.py#L2918, where XSoftmax's backward returns None.

Reply:
I'm confused, because when we trace the backward, always_restore should be True, so we should never get here

Reply:
Oh I see what's going on

Reply:
Proposal: the logic should be something like:
```
if always_restore:
   # we don't care about what the output is
   pass
else:
   # we require the output to be a list/tuple of Tensors
   # If it isn't then we raise notimplemented
```
   
When always_restore is True, we don't actually care what the output is, because we are never going to use it to construct a subgraph. I don't want to allow non-Tensor output in higher order ops just yet but it might be inevitable.